src/Microsoft.ComponentDetection.Detectors/dockerfile/DockerfileComponentDetector.cs

```diff
@@ -0,0 +1,154 @@
+    public class DockerfileComponentDetector : FileComponentDetector
+    {
+        [Import]
+        public ICommandLineInvocationService CommandLineInvocationService { get; set; }
+
+        [Import]
+        public IEnvironmentVariableService EnvVarService { get; set; }
+
+        public override string Id { get; } = "dockerfile";
+
+        public override IEnumerable<string> Categories => new[] { Enum.GetName(typeof(DetectorClass), DetectorClass.GoMod) };
+
+        public override IList<string> SearchPatterns { get; } = new List<string> { "dockerfile", "dockerfile.*" };
+
+        public override IEnumerable<ComponentType> SupportedComponentTypes { get; } = new[] { ComponentType.Dockerfile };
+
+        public override int Version => 5;
+
+        private HashSet<string> projectRoots = new HashSet<string>();
+
+        protected override async Task OnFileFound(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+        {
+            var singleFileComponentRecorder = processRequest.SingleFileComponentRecorder;
+            var file = processRequest.ComponentStream;
+            var filePath = file.Location;
+            Logger.LogInfo($"Discovered dockerfile: {file.Location}");
+
+            string contents;
+            using (var reader = new StreamReader(file.Stream))
+            {
+                contents = await reader.ReadToEndAsync();
+            }
+
+            var stageNameMap = new Dictionary<string, string>();
+            var dockerFileComponent = ParseDockerFile(contents, file.Location, singleFileComponentRecorder, stageNameMap);
+        }
+
+        private Task ParseDockerFile(string fileContents, string fileLocation, ISingleFileComponentRecorder singleFileComponentRecorder, Dictionary<string, string> stageNameMap)
+        {
+            var dockerfileModel = Valleysoft.DockerfileModel.Dockerfile.Parse(fileContents);
+            var instructions = dockerfileModel.Items;
+            foreach (var instruction in instructions)
+            {
+                var imageReference = ProcessDockerfileConstruct(instruction, dockerfileModel.EscapeChar, stageNameMap);
+                if (imageReference != null)
+                {
+                    singleFileComponentRecorder.RegisterUsage(new DetectedComponent(imageReference.ToTypedDockerImageComponent()));
+                }
+            }
+            
+            return Task.CompletedTask;
+        }
+
+        private IDockerReference ProcessDockerfileConstruct(DockerfileConstruct construct, char escapeChar, Dictionary<string, string> stageNameMap)
+        {
+            var instructionKeyword = construct.Type;
+            IDockerReference baseImage = null;
+            if (instructionKeyword == ConstructType.Instruction)
+            {
+                var constructType = construct.GetType().ToString();
+                switch (constructType)
+                {
+                    case "Valleysoft.DockerfileModel.FromInstruction":
```

Comment:
Magic strings? Is there an exported type you can use instead?

Reply:
There is, but Switch cases based on Type are in preview, and apparently not widely accepted: https://docs.microsoft.com/en-us/archive/blogs/jaredpar/switching-on-types